### ISSUE TITLE
Discussion: Migrate to new config method

### ISSUE THREAD

**Issue by jcwillox** [lptstr/winfetch#29]
I've created a demo version of my suggested changes at [jcwillox/winfetch@demo](https://github.com/jcwillox/winfetch/tree/demo). 
Specifically, this commit https://github.com/jcwillox/winfetch/commit/cb003ea73c1a2933437ab83e7a751af954e7b8d0 is where I make the changes. 
The demo branch is based upon the 3 PRs I recently created.

The approach I have used is far more modular than the current method, this should make it easier to add info segments and it allows users to add custom segments to their configs.
It also allows users to configure the order of segments and add blank lines (#16).
It simplifies the printing out loop, and prints info segments dynamically as the information is retrieved (like neofetch does).

The changes are in three parts, switching to the new config method (and converting old configs), encapsulating each info segment inside a function, and rewriting the printing loop for the new method.

TLDR; **Implications**:
* Change order of info segments
* Insert blank lines
* Can hide color bar
* Add custom info segments in config
* Info lines are written out progressively (users can see progress)
* Modular approach simplifies code and makes it easier to add info lines
* Fixes no-image (it is currently broken)

**Example Config**:
```powershell
# ===== WINFETCH CONFIGURATION =====

# Remove the '#' from any of the lines in
# this file to **enable** their output.

# $noimage = $true

# create custom info line
function info_custom_time {
    return @{
        title = "Time"
        content = (Get-Date)
    }
}

@(
    "title"
    "dashes"
    "os"
    "computer"
    "custom_time"  # use custom info line
    "uptime"
    "pkgs"
    "pwsh"
    "terminal"
    "cpu"
    "gpu"
    "memory"
    "disk"
    "blank"  # empty info line
    "colorbar"
)
```

This is, of course, open to discussion, if anyone has a better idea, styling suggestions, etc. Otherwise, if the maintainer is happy, I'll make some final tweaks and submit a PR, where we can finalize any changes.

**Comment by kiedtl** [comment]
Give me a day or so : it's been ages since I've used a Windows machine,
and I'm going to need to set up a VM to test this out...

**Comment by rashil2000** [comment]
Hi @jcwillox , Just wanted to give a couple of suggestions as a part of this discussion. 

- Adding the ability to truncate output depending on the terminal window's width, would be a good idea. Currently it jumps onto the next line and messes up the whole output. Neofetch already this, i.e. it prints only till the available number of columns. (You can get the current number of available columns in PowerShell using `$Host.UI.RawUI.WindowSize.Width`) 
- Some improvements for detecting windows terminals can be made to work with all versions of Powershell (like 5.1). I think [this](https://github.com/roose/winfetch/commit/1f9690b6e630d94e487aff10043853f42a01055f) is a good reference to get started...

**Comment by kiedtl** [comment]
Unfortunately, I'm going to have to ask for Another Few Days(tm). My low-end machine doesn't seem to enjoy VirtualBox very much...

**Comment by rashil2000** [comment]
@kiedtl Might I suggest adding some new maintainers/contributors who have access to Windows machines at all times or can actively participate?

**Comment by rashil2000** [comment]
@kiedtl Please don't take it in a negative way, I didn't mean to offend.

Personally I think winfetch is by far the most feature-rich PowerShell alternative to neofetch (considering alternatives like [screenfetch](https://github.com/JulianChow94/Windows-screenFetch) and [systemsplash](https://github.com/mmillar-bolis/SystemSplash)). I'd love to see more functionality, like the following, added to it to make it as good as (or even better than) neofetch. And I'm more than willing to contribute these additions myself :)

- Motherboard information 
- Multiple displays information 
- Multiple disks information 
- Process count
- Processor load
- Graph-like metrics
- Support for more Windows logos
- App theme or system theme detection 
- Battery status 
- Locale (culture) information 
- GPU Driver
- Songs, music player information 
- Local IP, Public IP address

While most of these are easy to do, some might not be possible. I'm open to suggestions. 

Thanks for making this awesome tool btw :)

**Comment by jcwillox** [comment]
@kiedtl that's all good, running Windows 10 in a VM is annoyingly difficult, I can't even run a Windows VM on Windows using VirtualBox because of WSL2.

@rashil2000 I've actually got the code ready for a couple of those (local/public IP, battery, motherboard), and it should be much easier to add new core and custom info lines with this config style or some variant of it, as the majority of the code is independent of the info functions.

**Comment by rashil2000** [comment]
@jcwillox Even I have the code for some of those ready (like theme detection, processes, load, disks etc.). Agreed that those functions are independent of the config method itself. 

Btw it's a limitation of VirtualBox itself, it can't run when Hyper-V is enabled. And WSL2 needs Hyper-V. I'd suggest using Qemu or VMware Player as both utilize Hyper-V now.

**Comment by jcwillox** [comment]
Update: 
I'm changing `[Console]::Write` to `Write-Host` as unlike `[Console]::Write` it can be redirected (using `6>&1`).

The mixed-use of `[Console]::Write` and `Write-Output` causes the console to get messed up when trying to redirect output or when using for example `Measure-Command`.

The only reason I was using `[Console]::Write` (or `Write-Host`) is because it allows me to keep the cursor on column 40, which I think looks better. When using `Write-Output` the cursor will on column 0.

Currently, I'm either going to change it to all use `Write-Host` or alternatively use `Write-Output` and temporarily hide the cursor using ANSI escapes. I haven't decided which is better?

**Comment by kiedtl** [comment]
> Might I suggest adding some new maintainers/contributors who have access to Windows machines at all times or can actively participate?

Unfortunately, I am not aware of anyone who is interesting in maintaining winfetch...

> Please don't take it in a negative way, I didn't mean to offend.

No worries, I've been interested in adding another maintainer for some time :^)

>  I'd love to see more functionality, like the following, added to it to make it as good as (or even better than) neofetch. And I'm more than willing to contribute these additions myself :)

Huh, that would be great! It would convert Winfetch to something more than a PowerShell clone of Neofetch.

> Currently, I'm either going to change it to all use Write-Host or alternatively use Write-Output and temporarily hide the cursor using ANSI escapes. I haven't decided which is better?

Hmm. To be honest, I don't really care too much about hiding the cursor; it's not a big deal. On the other hand, it might confuse users into thinking that Winfetch has completed(?)

**Comment by kiedtl** [comment]
> This is, of course, open to discussion, if anyone has a better idea, styling suggestions, etc. Otherwise, if the maintainer is happy, I'll make some final tweaks and submit a PR, where we can finalize any changes.

The new configuration method looks good to me; go ahead and submit a pull request, and I'll take a final review of the code. (Before this is done, I intend to make a minor release.)

**Comment by kiedtl** [comment]
@jcwillox @rashil2000

**Comment by rashil2000** [comment]
> Unfortunately, I am not aware of anyone who is interesting in maintaining winfetch...

@kiedtl I can willingly help in reviewing PRs/adding my own features etc. :)

**Comment by kiedtl** [comment]
> I can willingly help in reviewing PRs/pushing my own features etc. :)

Cool, I'll keep that in mind! In the mean time, I would definitely appreciate help with reviewing any incoming PRs ;)

**Comment by jcwillox** [comment]
I've finished my touch-ups and submitted the PR. I ended up going the `Write-Host` method, however, due to PowerShell's limitations if anyone wants to redirect the output to a file use `pwsh -Command "posh-winfetch.ps1" > output.txt` to get the correct formatting.

@kiedtl let me know if you require any functionality/style changes 👍.

Once this PR is merged I plan to submit subsequent PR's for adding `battery`, `motherboard`, `local_ip` and `public_ip` segments. I thought it would be best to break that up from the config migration PR.

**Comment by sheeepdev** [comment]
Hi y'all, I use winfetch most of the time. I'd love to see these features, mainly editing my config to lowercase, editing lines, and contribute to winfetch but I do not know how to code or how to use GitHub either or PRs and other things, just know they exist and (some) of how they function. Hope more features will be added to winfetch and that I'll be able to contribute in the future. Also, I do see that winfetch needs contributors so ill try learning all this :)
